pkg/utils/hwaddr/hwaddr.go

```diff
@@ -15,49 +15,27 @@
 package hwaddr
 
 import (
-	"fmt"
+	"math/rand"
 	"net"
 )
 
-const (
-	ipRelevantByteLen      = 4
-	PrivateMACPrefixString = "0a:58"
-)
-
-var (
-	// private mac prefix safe to use
-	PrivateMACPrefix = []byte{0x0a, 0x58}
-)
-
-type SupportIp4OnlyErr struct{ msg string }
-
-func (e SupportIp4OnlyErr) Error() string { return e.msg }
-
-type MacParseErr struct{ msg string }
-
-func (e MacParseErr) Error() string { return e.msg }
-
-type InvalidPrefixLengthErr struct{ msg string }
-
-func (e InvalidPrefixLengthErr) Error() string { return e.msg }
-
-// GenerateHardwareAddr4 generates 48 bit virtual mac addresses based on the IP4 input.
-func GenerateHardwareAddr4(ip net.IP, prefix []byte) (net.HardwareAddr, error) {
-	switch {
-
-	case ip.To4() == nil:
-		return nil, SupportIp4OnlyErr{msg: "GenerateHardwareAddr4 only supports valid IPv4 address as input"}
-
-	case len(prefix) != len(PrivateMACPrefix):
-		return nil, InvalidPrefixLengthErr{msg: fmt.Sprintf(
-			"Prefix has length %d instead  of %d", len(prefix), len(PrivateMACPrefix)),
-		}
-	}
-
-	ipByteLen := len(ip)
-	return (net.HardwareAddr)(
-		append(
-			prefix,
-			ip[ipByteLen-ipRelevantByteLen:ipByteLen]...),
-	), nil
+// The first 24 bits of the MAC represent the Organizationally Unique Identifier (OUI),
+// and the first byte of the MAC address has two special bits.
+// 1. The least-significant bit: 0 for unicast and 1 for multicast
+// 2. The second-least-significant bit: 0 for globally unique and 1 for locally administered
+// Since MAC address for container interfaces is locally administered and unicast, so we can
+// fix the two LSb of MSB to 10 and use any values for other bits. This will avoid any conflicts
+// with public OUIs.
+// We use 02:58:00 for CNI official OUI here.
+var cniOUI = []byte{0x02, 0x58, 0x00}
+
+// GenerateMAC will generate MAC addresses with fixed first 24 bits (CNI OUI), and the last 24 bits
+// will be random, so there are at most 16777216(2^24) different addresses.
+// To avoid MAC address collision, we suggest to use this function when length of subnet masking is
+// greater or equal than 22 in IPAM.
+func GenerateMAC() net.HardwareAddr {
+	hw := make(net.HardwareAddr, 6)
+	copy(hw[:3], cniOUI)
+	_, _ = rand.Read(hw[3:])
```

Comment:
rand.Seed() is not invoked before rand.Read(). Without 'rand.Seed()' call, rand.Read() uses 'rand.Seed(1)' as you see in [document](https://golang.org/pkg/math/rand/#Seed), hence it is not random and cracker could infer the next MAC address. Call rand.Seed() in plugin side can be the 
possible solution, but it may be missed, so how about to add rand.Seed() in init()?

Reply:
Thanks for the good catch! And @aojea has provided another way above, switching `math/rand` to `crypto/rand`, which will make MAC addresses more encrypted and secure. More important, not rely on `Seed` anymore.